exonum-java-binding/core/src/main/java/com/exonum/binding/core/runtime/TransactionMethodExtractor.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.core.runtime;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.exonum.binding.core.transaction.TransactionContext;
+import com.exonum.binding.core.transaction.TransactionMethod;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Finds and validates transaction methods in a service.
+ */
+final class TransactionMethodExtractor {
+
+  /**
+   * Returns a map of transaction ids to transaction methods found in a service class.
+   *
+   * @see TransactionMethod
+   */
+  static Map<Integer, MethodHandle> extractTransactionMethods(Class<?> serviceClass) {
+    Map<Integer, MethodHandle> transactions = new HashMap<>();
+    MethodHandles.Lookup lookup = MethodHandles.lookup();
+    while (serviceClass != Object.class) {
```

Comment:
This method does several things making testing more complicated:

   * Finds all methods annotated with a particular annotation
   * Validates these annotations
   * Converts to something callable

At least the first step, contributing to this outer loop, must be extracted and tested
      separately.